### ISSUE TITLE
Fix the HashMap Performance

### ISSUE THREAD

**Issue by StefanoD** [rust-lang/rust#34759]
See: https://users.rust-lang.org/t/hashmap-performance/6476

**Comment by arthurprs** [comment]
Here're are my results with latest nightly. **Simple** in this case is the same strategy used by unordered_map, just return the integer.  We're actually faster.

```
2000
SIP13 Duration { secs: 6, nanos: 709803720 }
2000
SIP24 Duration { secs: 7, nanos: 654518957 }
2000
FNV Duration { secs: 5, nanos: 179056737 }
2000
Simple Duration { secs: 0, nanos: 897325662 }
➜  temp ./a.out 
2000 map 18.505 s
2000 unordered 1.9361 s
```

**Comment by brson** [comment]
While I appreciate the sentiment, as stated I don't see this as being actionable. The topic has been revisited many times, and Rust's hashmaps are the way they are for reasons.

Some alternate ideas for issues to address this area: 'Add Fnv hashing to std', 'specialize HashMap for small keys' (with a plausible strategy).

**Comment by inv2004** [comment]
@brson , Sorry if I missed something, but the problem with HashMap performance is not the problem with hash function. I just found the issue and I found exactly the same problem: HashMap with hasher which use key like hash without any modifications is much slower is rust. You can compare with indexmap for example.

**Comment by arthurprs** [comment]
@inv2004 can you rephrase that? I'm not sure I follow.

**Comment by inv2004** [comment]
@arthurprs 
The most of the topic about performance problem with collections::HashMap is about hash-functions. But we can clearly see big difference in performance of indexmap and collections::HashMap with the same hash-function.

**Comment by arthurprs** [comment]
@inv2004 IndexMap makes different tradeoffs, it's not a slam-dunk like you're suggesting. While iteration is much faster, there's a trade-off towards insertion speed (away from lookup and memory). The load factor 75% (compared to 92%) plays a big factor.
